kroxylicious-runtime/src/main/templates/BodyDecoder.ftl

```diff
@@ -64,18 +63,39 @@ public class BodyDecoder {
     * @param accessor the accessor for the message bytes
     * @return the ApiMessage
     * @throws IllegalArgumentException if an unhandled ApiKey is encountered
+    * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-511%3A+Collect+and+Expose+Client%27s+Name+and+Version+in+the+Brokers#KIP511:CollectandExposeClient'sNameandVersionintheBrokers-ApiVersionsRequest/ResponseHandling">KIP-511: Collect and Expose Client's Name and Version in the Brokers</a>
+    * ApiVersions Request/Response Handling
     */
-    static ApiMessage decodeResponse(ApiKeys apiKey, short apiVersion, Readable accessor) {
-        switch (apiKey) {
+    static ApiMessage decodeResponse(ApiKeys apiKey, short apiVersion, ByteBufAccessor accessor) {
+        return switch (apiKey) {
 <#list messageSpecs as messageSpec>
     <#if messageSpec.type?lower_case == 'response'>
-            case ${retrieveApiKey(messageSpec)}:
-                return new ${messageSpec.name}Data(accessor, apiVersion);
+        <#if messageSpec.name == 'ApiVersionsResponse'>
+            case ${retrieveApiKey(messageSpec)} -> {
+                // KIP-511 when the client receives an unsupported version for the ApiVersionResponse, it fails back to version 0
+                // Use the same algorithm as https://github.com/apache/kafka/blob/a41c10fd49841381b5207c184a385622094ed440/clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java#L90-L106
+                int prev = accessor.readerIndex();
+                try {
+                    yield new ${messageSpec.name}Data(accessor, apiVersion);
```

Comment:
I don't understand the use of `yield` here. It looks like when the template is rendered it gets turned into a `return`

Reply:
That's not the case for me.   I see the yield in the generated `io.kroxylicious.proxy.internal.codec.BodyDecoder`.   Did your build perform the code-generation right?

Reply:
When I generate code on this branch i get 
```java
static ApiMessage decodeResponse(ApiKeys apiKey, short apiVersion, ByteBufAccessor accessor) {
        return switch (apiKey) {
            case ADD_OFFSETS_TO_TXN -> new AddOffsetsToTxnResponseData(accessor, apiVersion);
            case ADD_PARTITIONS_TO_TXN -> new AddPartitionsToTxnResponseData(accessor, apiVersion);
            case ALLOCATE_PRODUCER_IDS -> new AllocateProducerIdsResponseData(accessor, apiVersion);
            case ALTER_CLIENT_QUOTAS -> new AlterClientQuotasResponseData(accessor, apiVersion);
            case ALTER_CONFIGS -> new AlterConfigsResponseData(accessor, apiVersion);
            case ALTER_PARTITION_REASSIGNMENTS -> new AlterPartitionReassignmentsResponseData(accessor, apiVersion);
            case ALTER_PARTITION -> new AlterPartitionResponseData(accessor, apiVersion);
            case ALTER_REPLICA_LOG_DIRS -> new AlterReplicaLogDirsResponseData(accessor, apiVersion);
            case ALTER_USER_SCRAM_CREDENTIALS -> new AlterUserScramCredentialsResponseData(accessor, apiVersion);
            case API_VERSIONS -> {
                // KIP-511 when the client receives an unsupported version for the ApiVersionResponse, it fails back to version 0
                // Use the same algorithm as https://github.com/apache/kafka/blob/a41c10fd49841381b5207c184a385622094ed440/clients/src/main/java/org/apache/kafka/common/requests/ApiVersionsResponse.java#L90-L106
                int prev = accessor.readerIndex();
                try {
                    yield new ApiVersionsResponseData(accessor, apiVersion);
                }
                catch (RuntimeException e) {
                    accessor.readerIndex(prev);
                    if (apiVersion != 0) {
                        yield new ApiVersionsResponseData(accessor, (short) 0);
                    }
                    else {
                        throw e;
                    }
                }
            }
            case ASSIGN_REPLICAS_TO_DIRS -> new AssignReplicasToDirsResponseData(accessor, apiVersion);
            case BEGIN_QUORUM_EPOCH -> new BeginQuorumEpochResponseData(accessor, apiVersion);
            case BROKER_HEARTBEAT -> new BrokerHeartbeatResponseData(accessor, apiVersion);
            case BROKER_REGISTRATION -> new BrokerRegistrationResponseData(accessor, apiVersion);
            case CONSUMER_GROUP_DESCRIBE -> new ConsumerGroupDescribeResponseData(accessor, apiVersion);
            case CONSUMER_GROUP_HEARTBEAT -> new ConsumerGroupHeartbeatResponseData(accessor, apiVersion);
            case CONTROLLED_SHUTDOWN -> new ControlledShutdownResponseData(accessor, apiVersion);
            case CONTROLLER_REGISTRATION -> new ControllerRegistrationResponseData(accessor, apiVersion);
            case CREATE_ACLS -> new CreateAclsResponseData(accessor, apiVersion);
            case CREATE_DELEGATION_TOKEN -> new CreateDelegationTokenResponseData(accessor, apiVersion);
            case CREATE_PARTITIONS -> new CreatePartitionsResponseData(accessor, apiVersion);
            case CREATE_TOPICS -> new CreateTopicsResponseData(accessor, apiVersion);
            case DELETE_ACLS -> new DeleteAclsResponseData(accessor, apiVersion);
            case DELETE_GROUPS -> new DeleteGroupsResponseData(accessor, apiVersion);
            case DELETE_RECORDS -> new DeleteRecordsResponseData(accessor, apiVersion);
            case DELETE_TOPICS -> new DeleteTopicsResponseData(accessor, apiVersion);
            case DESCRIBE_ACLS -> new DescribeAclsResponseData(accessor, apiVersion);
            case DESCRIBE_CLIENT_QUOTAS -> new DescribeClientQuotasResponseData(accessor, apiVersion);
            case DESCRIBE_CLUSTER -> new DescribeClusterResponseData(accessor, apiVersion);
            case DESCRIBE_CONFIGS -> new DescribeConfigsResponseData(accessor, apiVersion);
            case DESCRIBE_DELEGATION_TOKEN -> new DescribeDelegationTokenResponseData(accessor, apiVersion);
            case DESCRIBE_GROUPS -> new DescribeGroupsResponseData(accessor, apiVersion);
            case DESCRIBE_LOG_DIRS -> new DescribeLogDirsResponseData(accessor, apiVersion);
            case DESCRIBE_PRODUCERS -> new DescribeProducersResponseData(accessor, apiVersion);
            case DESCRIBE_QUORUM -> new DescribeQuorumResponseData(accessor, apiVersion);
            case DESCRIBE_TRANSACTIONS -> new DescribeTransactionsResponseData(accessor, apiVersion);
            case DESCRIBE_USER_SCRAM_CREDENTIALS -> new DescribeUserScramCredentialsResponseData(accessor, apiVersion);
            case ELECT_LEADERS -> new ElectLeadersResponseData(accessor, apiVersion);
            case END_QUORUM_EPOCH -> new EndQuorumEpochResponseData(accessor, apiVersion);
            case END_TXN -> new EndTxnResponseData(accessor, apiVersion);
            case ENVELOPE -> new EnvelopeResponseData(accessor, apiVersion);
            case EXPIRE_DELEGATION_TOKEN -> new ExpireDelegationTokenResponseData(accessor, apiVersion);
            case FETCH -> new FetchResponseData(accessor, apiVersion);
            case FETCH_SNAPSHOT -> new FetchSnapshotResponseData(accessor, apiVersion);
            case FIND_COORDINATOR -> new FindCoordinatorResponseData(accessor, apiVersion);
            case GET_TELEMETRY_SUBSCRIPTIONS -> new GetTelemetrySubscriptionsResponseData(accessor, apiVersion);
            case HEARTBEAT -> new HeartbeatResponseData(accessor, apiVersion);
            case INCREMENTAL_ALTER_CONFIGS -> new IncrementalAlterConfigsResponseData(accessor, apiVersion);
            case INIT_PRODUCER_ID -> new InitProducerIdResponseData(accessor, apiVersion);
            case JOIN_GROUP -> new JoinGroupResponseData(accessor, apiVersion);
            case LEADER_AND_ISR -> new LeaderAndIsrResponseData(accessor, apiVersion);
            case LEAVE_GROUP -> new LeaveGroupResponseData(accessor, apiVersion);
            case LIST_CLIENT_METRICS_RESOURCES -> new ListClientMetricsResourcesResponseData(accessor, apiVersion);
            case LIST_GROUPS -> new ListGroupsResponseData(accessor, apiVersion);
            case LIST_OFFSETS -> new ListOffsetsResponseData(accessor, apiVersion);
            case LIST_PARTITION_REASSIGNMENTS -> new ListPartitionReassignmentsResponseData(accessor, apiVersion);
            case LIST_TRANSACTIONS -> new ListTransactionsResponseData(accessor, apiVersion);
            case METADATA -> new MetadataResponseData(accessor, apiVersion);
            case OFFSET_COMMIT -> new OffsetCommitResponseData(accessor, apiVersion);
            case OFFSET_DELETE -> new OffsetDeleteResponseData(accessor, apiVersion);
            case OFFSET_FETCH -> new OffsetFetchResponseData(accessor, apiVersion);
            case OFFSET_FOR_LEADER_EPOCH -> new OffsetForLeaderEpochResponseData(accessor, apiVersion);
            case PRODUCE -> new ProduceResponseData(accessor, apiVersion);
            case PUSH_TELEMETRY -> new PushTelemetryResponseData(accessor, apiVersion);
            case RENEW_DELEGATION_TOKEN -> new RenewDelegationTokenResponseData(accessor, apiVersion);
            case SASL_AUTHENTICATE -> new SaslAuthenticateResponseData(accessor, apiVersion);
            case SASL_HANDSHAKE -> new SaslHandshakeResponseData(accessor, apiVersion);
            case STOP_REPLICA -> new StopReplicaResponseData(accessor, apiVersion);
            case SYNC_GROUP -> new SyncGroupResponseData(accessor, apiVersion);
            case TXN_OFFSET_COMMIT -> new TxnOffsetCommitResponseData(accessor, apiVersion);
            case UNREGISTER_BROKER -> new UnregisterBrokerResponseData(accessor, apiVersion);
            case UPDATE_FEATURES -> new UpdateFeaturesResponseData(accessor, apiVersion);
            case UPDATE_METADATA -> new UpdateMetadataResponseData(accessor, apiVersion);
            case VOTE -> new VoteResponseData(accessor, apiVersion);
            case WRITE_TXN_MARKERS -> new WriteTxnMarkersResponseData(accessor, apiVersion);

            default -> throw new IllegalArgumentException("Unsupported RPC " + apiKey);
        };
```

Which looks right to me, given this is now a [switch expression](https://docs.oracle.com/en/java/javase/17/language/switch-expressions.html#GUID-BA4F63E3-4823-43C6-A5F3-BAA4A2EF3AD) rather than a switch statment

Reply:
Thanks @SamBarker, of course, that makes sense now.